File: attentive-android-sdk/src/androidTest/java/com/attentive/androidsdk/PersistentStorageTest.java
package com.attentive.androidsdk;

import static org.junit.Assert.assertEquals;

import android.content.Context;
import androidx.test.ext.junit.runners.AndroidJUnit4;
import androidx.test.platform.app.InstrumentationRegistry;
import org.junit.Test;
import org.junit.runner.RunWith;

@RunWith(AndroidJUnit4.class)
public class PersistentStorageTest {
    @Test
    public void saveAndRead_addsOneKeyValuePair_keyValuePairIsReturned() {
        // Arrange
        Context appContext = InstrumentationRegistry.getInstrumentation().getTargetContext();
        PersistentStorage persistentStorage = new PersistentStorage(appContext);
        final String key = "someKey";
        final String value = "someValue";

        // Act
        persistentStorage.save(key, value);

        // Assert
        assertEquals(value, persistentStorage.read(key));
    }

    @Test
    public void saveAndRead_overwriteExistingValue_readReturnsNewValue() {
        // Arrange
        Context appContext = InstrumentationRegistry.getInstrumentation().getTargetContext();
        PersistentStorage persistentStorage = new PersistentStorage(appContext);
        final String key = "someKey";
        final String firstValue = "someValue";
        final String secondValue = "someOtherValue";

        // Act
        persistentStorage.save(key, firstValue);
        persistentStorage.save(key, secondValue);

        // Assert
        assertEquals(secondValue, persistentStorage.read(key));
    }
}

File: attentive-android-sdk/src/androidTest/java/com/attentive/androidsdk/PersistentStorageTest.kt
package com.attentive.androidsdk

import androidx.test.ext.junit.runners.AndroidJUnit4
import androidx.test.platform.app.InstrumentationRegistry
import org.junit.Assert
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith

@RunWith(AndroidJUnit4::class)
class PersistentStorageTest {
    private var persistentStorage: PersistentStorage? = null

    @Before
    fun setup() {
        persistentStorage =
            PersistentStorage(InstrumentationRegistry.getInstrumentation().targetContext)
    }

    @Test
    fun saveAndRead_addsOneKeyValuePair_keyValuePairIsReturned() {
        // Arrange
        val key = "someKey"
        val value = "someValue"

        // Act
        persistentStorage!!.save(key, value)

        // Assert
        Assert.assertEquals(value, persistentStorage!!.read(key))
    }

    @Test
    fun saveAndRead_overwriteExistingValue_readReturnsNewValue() {
        // Arrange
        val key = "someKey"
        val firstValue = "someValue"
        val secondValue = "someOtherValue"

        // Act
        persistentStorage!!.save(key, firstValue)
        persistentStorage!!.save(key, secondValue)

        // Assert
        Assert.assertEquals(secondValue, persistentStorage!!.read(key))
    }

    @Test
    fun read_readKeyThatDoesNotExist_returnsNull() {
        // Arrange
        val key = "someKey"

        // Act
        persistentStorage!!.delete(key)

        // Assert
        Assert.assertNull(persistentStorage!!.read(key))
    }

    @Test
    fun delete_deleteKeyThatDoesNotExist_noop() {
        // Arrange
        val key = "someKey"

        // Act
        persistentStorage!!.delete(key)

        // Assert
        Assert.assertNull(persistentStorage!!.read(key))
    }

    @Test
    fun delete_deleteExistingKey_readReturnsNull() {
        // Arrange
        val key = "someKey"
        val value = "someValue"
        persistentStorage!!.save(key, value)

        // Act
        persistentStorage!!.delete(key)

        // Assert
        Assert.assertEquals(null, persistentStorage!!.read(key))
    }
}
